assembly/src/assembler/instruction/crypto_ops.rs

```diff
@@ -1,9 +1,55 @@
-use super::{AssemblyError, CodeBlock, Operation::*, SpanBuilder};
+use super::{AssemblyError, CodeBlock, Operation::*, SpanBuilder, ONE};
 use vm_core::{AdviceInjector, Decorator};
 
 // HASHING
 // ================================================================================================
 
+/// Appends HPERM and stack manipulation operations to the span block as required to compute a
+/// 1-to-1 hash. The top of the stack is expected to be arranged with 1 word (4 elements) to be
+/// hashed: [A, ...]. The resulting stack will contain the 1-to-1 hash result: [C, ...].
+///
+/// This assembly instruction uses the VM operation HPERM at its core, which permutes the top 12
+/// elements of the stack.
+///
+/// To perform the operation we do the following:
+/// 1. Prepare the stack with 12 elements for HPERM by pushing 4 more elements for the capacity,
+///    then reordering the stack and pushing an additional 4 elements so that the stack looks
+///    like: [1, 0, 0, 0, a3, a2, a1, a0, 1, 0, 0, 0, ...].  The first capacity element is set to
+///    ONE as we are hashing a number of elements which is not a multiple of the rate width. We
+///    also set the next element in the rate after `A` to ONE.  All other capacity and rate
+///    elements are set to ZERO, in accordance with the RPO rules.
+/// 2. Append the HPERM operation, which performs a permutation of RPO on the top 12 elements and
+///    leaves the an output of [D, C, B, ...] on the stack.  C is our 1-to-1 has result.
+/// 3. Drop D and B to achieve our result [C, ...]
+///
+/// This operation takes 19 VM cycles.
+pub(super) fn hash(span: &mut SpanBuilder) -> Result<Option<CodeBlock>, AssemblyError> {
+    #[rustfmt::skip]
+    let ops = [
+        // add 4 elements to the stack to be used as the capacity elements for the RPO permutation
+        Push(ONE), Pad, Pad, Pad,
+
+        // swap capacity elements such that they are below the elements to be hashed
+        SwapW,
+
+        // add 4 additional elements to the stack to populate the remaining elements of the rate
+        Push(ONE), Pad, Pad, Pad,
+
+        // Apply a hashing permutation on the top 12 elements in the stack
+        HPerm,
+
+        // Drop 4 elements (the part of the rate that doesn't have our result)
+        Drop, Drop, Drop, Drop,
+
+        // Move the top word (our result) down the stack
+        SwapW,
+
+        // Drop 4 elements (the capacity portion)
+        Drop, Drop, Drop, Drop,
```

Comment:
We usually try to replace `push.1` with `PAD INCR`. It takes one extra cycle but also saves an operation group (an immediate value takes up a whole operation group). Here, we can replace line 36 with `Dup7, Dup7, Dup7, Dup7` and that would eliminate one `push.1` without adding an extra cycle.

We could also replace the first `PUSH(ONE)` with `PAD INCR` - but that does add one cycle bringing the total to 20. But I'm thinking it is still OK to do this.

Reply:
Good to know about the cost of immediate values, thanks for the explanation.

Is `Dup7` a single cycle?  Docs says `dup.n - (1-3 cycles)`?

Reply:
Yep, `Dup7` is a single cycle - see [here](https://github.com/0xPolygonMiden/miden-vm/blob/next/assembly/src/assembler/instruction/mod.rs#L183).

Reply:
right got it.  So I guess technically speaking the `Pad Incr` pattern is only more efficient in cases where the `Push(1)` would result in a new batch (current batch is saturated)?  However I guess for practical purposes it's probably best to just stick to the standard of using `Pad Incr` consistently?

Reply:
Yes, exactly. The assumption is that once these are inlined in larger programs, more frequently than not,  `Push(1)`  would result in more batches than `Pad Incr`.